adapters/kayzen/kayzen.go

```diff
@@ -0,0 +1,154 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpoint: *template,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) (requests []*adapters.RequestData, errors []error) {
+	var kayzenExt *openrtb_ext.ExtKayzen
+	var err error
+
+	if len(request.Imp) > 0 {
+		kayzenExt, err = a.getImpressionExt(&(request.Imp[0]))
+		if err != nil {
+			errors = append(errors, err)
+		}
+		request.Imp[0].Ext = nil
+	} else {
+		errors = append(errors, &errortypes.BadInput{
+			Message: "Missing Imp Object",
+		})
+	}
+
+	if len(errors) > 0 {
+		return nil, errors
+	}
+
+	url, err := a.buildEndpointURL(kayzenExt)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	return []*adapters.RequestData{{
+		Method:  http.MethodPost,
+		Body:    requestJSON,
+		Uri:     url,
+		Headers: headers,
+	}}, nil
+}
+
+func (a *adapter) getImpressionExt(imp *openrtb2.Imp) (*openrtb_ext.ExtKayzen, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Bidder extension not provided or can't be unmarshalled",
+		}
+	}
+	var kayzenExt openrtb_ext.ExtKayzen
+	if err := json.Unmarshal(bidderExt.Bidder, &kayzenExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+	return &kayzenExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtKayzen) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{
+		ZoneID:    params.Zone,
+		AccountID: params.Exchange,
+	}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", response.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "Bad Server Response",
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
```

Comment:
Prebid server core assigns `"USD"` as default currency if the bidResponse's `Currency` field is found to have an empty value. (Line 189 of `exchange/bidder.go` below)
```
187             if bidResponse != nil {
188                 // Setup default currency as `USD` is not set in bid request nor bid response
189                 if bidResponse.Currency == "" {
190                     bidResponse.Currency = defaultCurrency
191                 }
exchange/bidder.go
```
Will your bid server always bid in `"USD"`? If not, some adapters that use multi-currencies explicitly assign a currency value in their `MakeBids()` implementation. This is not a requirement is only a suggestion in case it makes sense in the context of your adapter bidder.
```
100   func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
101       if response.StatusCode == http.StatusNoContent {
102           return nil, nil
103       }
104   *-- 21 lines: if response.StatusCode == http.StatusBadRequest {--------------------------------------------------------------------------------------------------------------------------------------------------------------------
125  
126       bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
    +     bidResponse.Currency = response.Cur
127  
128       var errs []error
129       for _, seatBid := range bidResp.SeatBid {
130   *--  7 lines: for _, bid := range seatBid.Bid {------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
137       }
138       return bidResponse, errs
139   }
adapters/kayzen/kayzen.go
```

Reply:
Thanks for raising this
Yes our bidder always bid in USD, So that should not be an issue

Reply:
Thank you for addressing our feedback.